pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -791,53 +770,401 @@ func (r *MachineSetSyncReconciler) createOrUpdateCAPIInfraMachineTemplate(ctx co
+	if err != nil {
+		return false, fmt.Errorf("failed to check if patch is required: %w", err)
+	}
+
+	if isPatchRequired {
+		if err := r.Status().Patch(ctx, existingMAPIMachineSet, patchBase); err != nil {
+			logger.Error(err, "Failed to update MAPI machine set status")
+
+			updateErr := fmt.Errorf("failed to update MAPI machine set status: %w", err)
+
+			if condErr := r.applySynchronizedConditionWithPatch(
+				ctx, existingMAPIMachineSet, corev1.ConditionFalse, reasonFailedToUpdateMAPIMachineSet, updateErr.Error(), nil); condErr != nil {
+				return false, utilerrors.NewAggregate([]error{updateErr, condErr})
+			}
+
+			return false, updateErr
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// setChangedCAPIMachineSetStatusFields sets the updated fields in the CAPI machine set status.
+//
+//nolint:gocognit,funlen
+func setChangedCAPIMachineSetStatusFields(existingCAPIMachineSet, convertedCAPIMachineSet *clusterv1.MachineSet) {
+	// convertedCAPIMachineSet holds the computed and desired status changes, so apply them to the existing existingCAPIMachineSet.
+	existingCAPIMachineSet.Status.Replicas = convertedCAPIMachineSet.Status.Replicas
+	existingCAPIMachineSet.Status.ReadyReplicas = convertedCAPIMachineSet.Status.ReadyReplicas
+	existingCAPIMachineSet.Status.AvailableReplicas = convertedCAPIMachineSet.Status.AvailableReplicas
+	existingCAPIMachineSet.Status.FullyLabeledReplicas = convertedCAPIMachineSet.Status.FullyLabeledReplicas
+	existingCAPIMachineSet.Status.FailureReason = convertedCAPIMachineSet.Status.FailureReason
+	existingCAPIMachineSet.Status.FailureMessage = convertedCAPIMachineSet.Status.FailureMessage
+
+	// Update the conditions if they exist (ignoring lasttransitiontime), append new ones.
+	for i := range convertedCAPIMachineSet.Status.Conditions { //nolint:dupl
+		hasCondition := false
+
+		for j := range existingCAPIMachineSet.Status.Conditions {
+			if existingCAPIMachineSet.Status.Conditions[j].Type == convertedCAPIMachineSet.Status.Conditions[i].Type {
+				hasChanged := false
+				hasCondition = true
+
+				if existingCAPIMachineSet.Status.Conditions[j].Status != convertedCAPIMachineSet.Status.Conditions[i].Status {
+					existingCAPIMachineSet.Status.Conditions[j].Status = convertedCAPIMachineSet.Status.Conditions[i].Status
+					hasChanged = true
+				}
+
+				if existingCAPIMachineSet.Status.Conditions[j].Reason != convertedCAPIMachineSet.Status.Conditions[i].Reason {
+					existingCAPIMachineSet.Status.Conditions[j].Reason = convertedCAPIMachineSet.Status.Conditions[i].Reason
+					hasChanged = true
+				}
+
+				if existingCAPIMachineSet.Status.Conditions[j].Message != convertedCAPIMachineSet.Status.Conditions[i].Message {
+					existingCAPIMachineSet.Status.Conditions[j].Message = convertedCAPIMachineSet.Status.Conditions[i].Message
+					hasChanged = true
+				}
+
+				if hasChanged {
+					// Set the last transition time to the new condition's last transition time only if the condition has changed.
```

Comment:
Should only change when the status changes normally, a transition is based only on the status, not on reason or message normally

Can you check what MAPI/CAPI are doing on their respective conditions?

Reply:
I did already look into this when I implemented this.

In CAPI (for both v1beta1 and the new v1beta2), a condition is considered to have changed (and its LastTransitionTime is updated) when ANY of the fields changes (so Status, Reason, Message, ...) and not just when the Status changes (see [here](https://github.com/kubernetes-sigs/cluster-api/blob/26c8b19a6ba039542d32a1d16dd40d17ef6d0fe6/util/conditions/patch.go#L237-L254))

In MAPI same story (see [here](https://github.com/openshift/machine-api-operator/blob/7c2a7004121df71cbec5802e4965876dcd34f5d0/pkg/util/conditions/gettersetter.go#L174-L182))

Reply:
Source: https://github.com/kubernetes-sigs/cluster-api/blob/main/util/conditions/setter.go#L60-L64